qbraid_qir/qasm3/visitor.py

```diff
@@ -165,6 +172,12 @@ def _check_in_scope(self, var_name: str) -> bool:
         curr_scope = self._get_scope()
         return var_name in curr_scope
 
+    def _find_in_visible_scope(self, var_name: str) -> Union[Variable, None]:
```

Comment:
I get that a simple scope check may not be the right way to determine whether a variable is declared. But, the scopes which are pushed may not need to be visible in all cases. Eg. - 

```c 
OPENQASM 3.0;
qubit[3] q;

def func2() { 
   rx(pi * a) q[1];
}

def func1 () {
    int[32] a = 4;
    x q[0];
    func2();
}

func1();
```
In this case, although `a` would not be defined in `func2` (but available in the constructed scope of the `func1`), according to your scope check it would be found in the "visible scope" which is incorrect. 

You would need to check the scope of the parent of the loop (or maybe even further) but this needs to be limited to either the function scope (in which this loop executes) or the global scope.

Reply:
Thanks for the feedback! I realised this at the time, but as far as I can see the implementation of context and scope currently present makes checking for this impossible (the contexts of previous scopes are not retained). The implementation in OpenQASM's own parser nests the scopes inside contexts and pushes a new context with its own set of scopes when the context changes - so that approach could be applicable here if that's not too large a change to the core logic of the visitor class. 

I initially refrained from attempting this because functions are currently unsupported, so the example above isn't something we can currently do, and it looks like it would be part of #70, which would require this? Sorry if I'm misunderstanding!

Reply:
I agree, it is a fair point that the subroutines aren't merged yet and the above case may not occur at the moment. 

The implementation would require some change about what **context** we are in. We could have a context indicating whether we are in a **loop**, **function**, or the **global scope**. So if we don't find the variable in the current scope AND we are in a loop context, we keep checking scopes up to the point where we reach either a function or global context.

We can keep this implementation for the hackathon's scope but it would be nice if you add a TODO comment over this implementation and reference this discussion.

Reply:
Thanks - I'll add the TODO for now but I'm interested to have another look at a later point :)